src/Data/Tree/DUAL/Internal.hs

```diff
@@ -146,178 +119,137 @@ pullU t@(Annot _ (DUALTreeU (u, _))) = pack (u, t)
-applyUpost u t = t <> leafU u
+leafU u = DUALTree u (Concat mempty)
 
 -- | Add an internal data value at the root of a tree.  Note that this
 --   only works on /non-empty/ trees; on empty trees this function is
---   the identity.
-annot :: (Semigroup u, Action d u) => a -> DUALTree d u a l -> DUALTree d u a l
-annot a = (over DUALTree . fmap) (pullU . Annot a)
+--   the identity. O(1)
+annot :: a -> DUALTree d u a l -> DUALTree d u a l
+annot _ EmptyDUAL      = EmptyDUAL
+annot a (DUALTree u t) = DUALTree u (Annot a t)
 
 -- | Apply a @d@ annotation at the root of a tree, transforming all
 --   @u@ annotations by the action of @d@.
-applyD :: (Semigroup d, Semigroup u, Action d u)
-       => d -> DUALTree d u a l -> DUALTree d u a l
-applyD = act . DAct
-
--- | Decompose a DUAL-tree into either @Nothing@ (if empty) or a
---   top-level cached @u@ annotation paired with a non-empty
---   DUAL-tree.
-nonEmpty :: DUALTree d u a l -> Maybe (u, DUALTreeNE d u a l)
-nonEmpty = fmap unpack . getOption . unpack
-
--- | Get the @u@ annotation at the root, or @Nothing@ if the tree is
---   empty.
-getU :: DUALTree d u a l -> Maybe u
-getU = fmap fst . nonEmpty
+down :: (Semigroup d, Semigroup u, Action d u) => d -> DUALTree d u a l -> DUALTree d u a l
+down _ EmptyDUAL      = EmptyDUAL
+down d (DUALTree u t) = DUALTree (act d u) $ case t of
+  Down d' t' -> Down (d <> d') t'
+  _          -> Down d t
 
-------------------------------------------------------------
--- Maps
-------------------------------------------------------------
+-- | Get the up annotation of a non-empty DUALTree.
+getU :: DUALTree d u a l -> Maybe u
+getU (DUALTree u _) = Just u
+getU _              = Nothing
 
--- XXX todo: try adding Map as a constructor, so we can delay the
--- mapping until the end too?
-
--- | Map a function (which must be a monoid homomorphism, and commute
---   with the action of @d@) over all the @u@ annotations in a non-empty
---   DUAL-tree.
-mapUNE :: (u -> u') -> DUALTreeNE d u a l -> DUALTreeNE d u' a l
-mapUNE f (Leaf u l)  = Leaf (f u) l
-mapUNE f (LeafU u)   = LeafU (f u)
-mapUNE f (Concat ts) = Concat ((NEL.map . mapUU) f ts)
-mapUNE f (Act d t)   = Act d (mapUU f t)
-mapUNE f (Annot a t) = Annot a (mapUU f t)
-
--- | Map a function (which must be a monoid homomorphism, and commute
---   with the action of @d@) over all the @u@ annotations in a
---   non-empty DUAL-tree paired with its cached @u@ value.
-mapUU :: (u -> u') -> DUALTreeU d u a l -> DUALTreeU d u' a l
-mapUU f = over DUALTreeU (f *** mapUNE f)
-
--- | Map a function over all the @u@ annotations in a DUAL-tree.  The
---   function must be a monoid homomorphism, and must commute with the
---   action of @d@ on @u@.  That is, to use @mapU f@ safely it must be
---   the case that
---
---     * @f mempty == mempty@
---
---     * @f (u1 \<\> u2) == f u1 \<\> f u2@
---
---     * @f (act d u) == act d (f u)@
---
```

Comment:
I don't think they were ever *required* from an implementation standpoint.  The point is that it leads to semantically strange things if you don't require these laws.  For example, I would hope that

```
mapU f d1 <> mapU f d2 === mapU f (d1 <> d2)
```

which doesn't hold in general unless you require those laws.  Maybe we should add the comment back but rewrite it to say something a bit more open-ended, like "if you want these nice properties to hold, it suffices to require that `f` have such-and-such properties; otherwise, use at your own risk"?